flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,165 @@
+    Preconditions.checkArgument(scanContext.endSnapshotId() == null,
+        "Cannot set end-snapshot-id option for streaming reader");
+    this.tableLoader = tableLoader;
+    this.scanContext = scanContext;
+  }
+
+  @Override
+  public void initializeState(FunctionInitializationContext context) throws Exception {
+    snapshotIdState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>(
+            "snapshot-id-state",
+            LongSerializer.INSTANCE));
+    tableLoader.open();
+    table = tableLoader.loadTable();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {}.", getClass().getSimpleName());
+      startSnapshotId = snapshotIdState.get().iterator().next();
+    } else {
+      Long optionStartSnapshot = scanContext.startSnapshotId();
+      if (optionStartSnapshot != null) {
+        if (!SnapshotUtil.ancestorOf(table, table.currentSnapshot().snapshotId(), optionStartSnapshot)) {
+          throw new IllegalStateException("The option start-snapshot-id " + optionStartSnapshot +
+              " is not an ancestor of the current snapshot");
+        }
+
+        startSnapshotId = optionStartSnapshot;
+      } else {
+        startSnapshotId = DUMMY_START_SNAPSHOT_ID;
+      }
+    }
+  }
+
+  @Override
+  public void snapshotState(FunctionSnapshotContext context) throws Exception {
+    snapshotIdState.clear();
+    snapshotIdState.add(startSnapshotId);
+  }
+
+  @Override
+  public void run(SourceContext<FlinkInputSplit> ctx) throws Exception {
+    checkpointLock = ctx.getCheckpointLock();
+    while (isRunning) {
+      synchronized (checkpointLock) {
+        if (isRunning) {
+          monitorAndForwardSplits(ctx);
+        }
+      }
+      Thread.sleep(scanContext.monitorInterval().toMillis());
+    }
+  }
+
+  private void monitorAndForwardSplits(SourceContext<FlinkInputSplit> ctx) {
+    table.refresh();
+    Snapshot snapshot = table.currentSnapshot();
+    if (snapshot != null && snapshot.snapshotId() != startSnapshotId) {
+      long snapshotId = snapshot.snapshotId();
+      // Read current static table if startSnapshotId not set.
+      ScanContext newScanContext = startSnapshotId == DUMMY_START_SNAPSHOT_ID ?
+          scanContext.copyWithSnapshotId(snapshotId) :
+          scanContext.copyWithAppendsBetween(startSnapshotId, snapshotId);
+
+      FlinkInputSplit[] splits = FlinkSplitGenerator.createInputSplits(table, newScanContext);
+      for (FlinkInputSplit split : splits) {
+        ctx.collect(split);
+      }
+      startSnapshotId = snapshotId;
+    }
+  }
+
+  @Override
+  public void cancel() {
+    // this is to cover the case where cancel() is called before the run()
+    if (checkpointLock != null) {
+      synchronized (checkpointLock) {
+        isRunning = false;
+      }
+    } else {
+      isRunning = false;
+    }
+
+    if (tableLoader != null) {
+      try {
+        tableLoader.close();
```

Comment:
Should this be called in `dispose` instead? The doc you pointed to says that dispose is when resources should be released, and that close may not be called:

> In the case of a termination due to a failure or due to manual cancellation, the execution jumps directly to the dispose() and skips any intermediate phases between the phase the operator was in when the failure happened and the dispose().

Reply:
Yes, you're right.  It's recommended to release all resources in an `AbstractStreamOperator`'s `dispose` method.  Here the `StreamingMonitorFuntion` is a  `RichSourceFunction` which does not have a `dispose` method.  But I've checked the  [AbstractUdfStreamOperator](https://github.com/apache/flink/blob/97bfd049951f8d52a2e0aed14265074c4255ead0/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractUdfStreamOperator.java#L117) (which is inherited by the `StreamSource` operator) ,  it's dispose method will call the function's `close` method directly.  So I think it's OK to release those resources here.